### PR TITLE
chore(*) introduce PaginationStore

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

Introduce the PaginationStore and delegate all pagination to be done there instead of implementation specific.

### CODE REVIEW

File: pkg/core/resources/store/pagination_store.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/kumahq/kuma/pkg/core/resources/registry"
+)
+
+func NewPaginationStore(delegate ResourceStore) ResourceStore {
+	return &PaginationStore{
+		delegate: delegate,
+	}
+}
+
+type PaginationStore struct {
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:17]: I think it can be private - `paginationStore`

Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:17]: Please put a doc why we are doing pagination in the memory so we don't forget

File: pkg/core/resources/store/pagination_store.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	for i := offset; i < offset+pageSize && i < lenAllItems; i++ {
+		for i < lenAllItems && !opts.Filter(allItems[i]) {
+			i++
+		}
+		if i == lenAllItems {
+			// no filtered items to add
+			break
+		}
+		_ = list.AddItem(allItems[i])
+	}
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:76]: can this be written as?
```
for i := offset; i < offset+pageSize && i < lenAllItems; i++ {
  if opts.Filter(allItems[i]) {
    _ = list.AddItem(allItems[i])
  }
}
```

File: pkg/api-server/dataplane_overview_endpoints.go [hunk 3]
```diff
@@ -173,3 +165,60 @@ func modeFromParameter(request *restful.Request, param string) (string, error) {
...
+			}
+
+			if !dataplane.Spec.MatchTags(tags) {
+				return false
+			}
+
+			return true
+		}
+	}, nil
+}
```
Comment (jakubdyszkiewicz) [inline pkg/api-server/dataplane_overview_endpoints.go:224]: Introduce
```
func ListByFilterFunc(filterFunc ListFilterFunc) ListOptionsFunc {
	return func(opts *ListOptions) {
		opts.FilterFunc = filterFunc
	}
}
```

here change the signature to
```
func genFilter(request *restful.Request) (store.ListFilterFunc, error) {
```

and then
```
func (r *dataplaneOverviewEndpoints) fetchOverviews(ctx context.Context, p page, meshName string, filter store.ListFilterFunc) (mesh.DataplaneOverviewResourceList, error) {
	dataplanes := mesh.DataplaneResourceList{}
	if err := r.resManager.List(ctx, &dataplanes, store.ListByMesh(meshName), store.ListByPage(p.size, p.offset), store.ListByFilterFunc(filter)); err != nil {
```

File: pkg/core/resources/store/options.go [hunk 4]
```diff
@@ -180,6 +183,15 @@ func NewListOptions(fs ...ListOptionsFunc) *ListOptions {
 	return opts
 }
 
+// Filter return true if the item passes the filtering criteria
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/options.go:186]: returns

File: pkg/core/resources/store/pagination_store.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+func (p *PaginationStore) Delete(ctx context.Context, resource model.Resource, optionsFunc ...DeleteOptionsFunc) error {
+	return p.delegate.Delete(ctx, resource, optionsFunc...)
+}
+
+func (p *PaginationStore) Get(ctx context.Context, resource model.Resource, optionsFunc ...GetOptionsFunc) error {
+	return p.delegate.Get(ctx, resource, optionsFunc...)
+}
+
+func (p *PaginationStore) List(ctx context.Context, list model.ResourceList, optionsFunc ...ListOptionsFunc) error {
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:37]: Optimisation:
if the `FilterFunc` is nil and `PageSize` is 0 and `PageOffset` is `""` just use delegate to original store.

For the most cases in our code we don't need pagination and filtering

File: pkg/core/resources/store/pagination_store.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	allItems := fullList.GetItems()
+	lenAllItems := len(allItems)
+
+	opts := NewListOptions(optionsFunc...)
+
+	offset := 0
+	pageSize := lenAllItems
+	paginateResults := opts.PageSize != 0
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:55]: `paginateResults`? How about `paginationEnabled` or something like  this

File: pkg/core/resources/store/pagination_store.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+	if paginateResults {
+		nextOffset := ""
+		if offset+pageSize < lenAllItems { // set new offset only if we did not reach the end of the collection
+			nextOffset = strconv.Itoa(offset + opts.PageSize)
+		}
+		list.GetPagination().SetNextOffset(nextOffset)
+	}
+
+	list.GetPagination().SetTotal(uint32(lenAllItems))
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:86]: I don't think it's right. I think we should filter first on the whole list, then do the pagination and the number here should be a result of the first filtering

**Review by jakubdyszkiewicz — CHANGES_REQUESTED** [review]

File: pkg/core/resources/store/pagination_store.go (hunk 6 above)
Comment (nickolaev) [reply pkg/core/resources/store/pagination_store.go:55]: FYI the code for pagination was moved here from the memory store.

**Comment by bloqhead** [comment]
I'm trying this out and this is the query:

```
http://localhost:5681/dataplanes+insights?gateway=true&size=12
```

I still receive this error though:

```
{
  "title": "Could not retrieve dataplane overviews",
  "details": "Pagination and filtering at the same time is not supported"
}
```
Please let me know if I'm doing something wrong 😅.Should this change work for Dataplane Insights also? My scenario is that I want to fetch only Gateway Dataplanes but I also want to reduce the size for pagination to 12 items per page. I would then do the same for Ingress and Standard (neither Ingress nor Gateway) Dataplanes.

I'm trying this in Universal on macOS. Thanks for the work you're putting in on this! It will make the GUI much more useful to the end user and give us a more robust HTTP API.

File: pkg/core/resources/store/pagination_store.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		return err
+	}
+
+	for _, item := range fullList.GetItems() {
+		if opts.Filter(item) {
+			_ = filteredList.AddItem(item)
+		}
+	}
+
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:70]: after filtering you need to do the sort to have stable pagination
```
sort.Sort(core_model.ByMeta(filteredList))
```

File: pkg/core/resources/store/pagination_store.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	filteredItems := filteredList.GetItems()
+	lenFilteresItems := len(filteredItems)
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/store/pagination_store.go:72]: `filteres` typo

**Review by jakubdyszkiewicz — CHANGES_REQUESTED** [review]

File: pkg/core/resources/store/pagination_store.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	"context"
+	"strconv"
+
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/registry"
+)
+
+// The Pagination Store is handling only the pagination functionality in the List.
+// This is an in-memory operation and offloads this from the persistent stores (k8s, postgres etc.)
+// which makes their implementation more simple. The in-memory filtering has been tested with 10,000
```
[the remark this answers is not in the document]
Comment (nickolaev) [reply pkg/core/resources/store/pagination_store.go:13]: OK, I don't see this to be an incorrect statement. We are not explaining what is and what is not a dealbreaker in the code IMO. The point is to leave a note here for our future selves what actually is going on here. When the next persistent storage gets introduced we'll not even consider paging, we'll just do the simple "get all" version and let this implementation deal with paging. Which is actually the point.

[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/core/resources/store/pagination_store.go:14]: I like to write comments that answer `why is this` (and what was the alternative) not `what is this`.

> The point is to leave a note here for our future selves what actually is going on here

Exactly. So imagine there is a new developer on the team and we have performance problems with pagination. He takes this task and finds this code and reads "which makes their implementation more simple" and "proved to be fast enough, although not that efficient". So I see there was a compromise, we thought it was fast enough but it isn't now. But what was the alternative? I'd like to see in the comment that
* There is no filtering + pagination on the native K8S database, so you have to do this in memory anyways.
* On Postgres, we keep the object in a column as a string. We would have to use JSON column type and convert it to native SQL queries.

> the next persistent storage gets introduced we'll not even consider paging  (...) Which is actually the point

IMHO the main point was to get Kube and Postgres to work

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**